File: 02_Optimize/gradient_descent.py
import numpy as np

def main():
    T = np.matrix("1; 2; 3; 4; 5")
    Y = np.matrix("0.8; 2.1; 3; 4.1; 5")
    Yh = lambda p: p[0] * T + p[1]
    P_init = np.array([2, -1])

    print(Y)
    print(Yh(P_init))
    print(jacobian(Yh, P_init))

    opt = Optimizer(Y, Yh, P_init)
    opt.set_step(0.01)
    opt.set_iteration(10000)
    P = opt.optimize()
    print(P)

class Optimizer:
    def __init__(self, Y, Yh, P):
        self.Y = Y   # Vector
        self.Yh = Yh # Function
        self.P = P   # Vector

    # Setter
    def set_step(self, alpha):
        self.alpha = alpha

    def set_iteration(self, num):
        self.num = num

    # Getter
    def get_sse(self):
        error = (self.Y - self.Yh(self.P))
        return (error.T * error)[0,0]

    def get_param(self):
        return self.P

    def update(self):
        J = jacobian(self.Yh, self.P)
        h = 2 * self.alpha * J * (self.Y - self.Yh(self.P))
        h = np.squeeze(np.array(np.squeeze(h)))
        self.P = self.P + h

    def optimize(self):
        for i in range(self.num):
            self.update()
            print(self.get_sse())
        return self.get_param()

def jacobian(Yh, P):
    m = len(P)
    n = len(Yh(P))
    h = np.zeros(m)
    J = np.zeros((m, n))
    for i in range(m):
        h_i = h
        h_i[i] = 1e-6
        P_i = P + h_i
        Yh_i = Yh(P_i) - Yh(P)
        for j in range(n):
            J[i, j] = Yh_i[j] / 1e-6
    return J

if __name__ == "__main__":
    main()

File: 02_Optimize/gradient_descent.jl
using ForwardDiff: jacobian

function main()
    T = Float64[1,2,3,4,5]
    Y = Float64[0.8, 2.1, 3, 4.1, 5]
    Yh = p -> @. p[1] * T + p[2]
    P_init = Float64[2,-1]
    println(Y)
    println(Yh(P_init))
    println(jacobian(Yh, P_init))

    opt = Optimizer(Y, Yh, P_init)
    P = opt |> optimize
    println(P)
end

# Optimizer
mutable struct Optimizer{T}
    Y::Vector{T}
    Yh::Function
    P::Vector{T}
    α::T
    num::Int
end

# Constructor
function Optimizer(Y, Yh, P, α=1e-2, num=10000)
    new(Optimizer(Y, Yh, P, α, num))
end

function set_step(opt::Optimizer{T}, α) where T <: Number
    opt.α = α
end

function set_iteration(opt::Optimizer{T}, num) where T <: Number
    opt.num = num
end

function get_sse(opt::Optimizer)
    error = opt.Y - opt.Yh(opt.P)
    (error' * error)[1,1]
end

function get_param(opt::Optimizer)
    opt.P
end

function update(opt::Optimizer)
    J = jacobian(opt.Yh, opt.P)
    h = 2 * opt.α * J' * (opt.Y - opt.Yh(opt.P))
    opt.P = opt.P .+ h
end

function optimize(opt::Optimizer)
    for i in 1:opt.num
        opt |> update
        opt |> get_sse |> println
    end
    opt |> get_param
end

main()

